File: frontend/src/providers/OpenAIProvider.ts
import OpenAI, {
  APIConnectionError,
  APIError,
  AuthenticationError,
  RateLimitError,
} from "openai";

import { DEFAULT_GENERATION_OPTIONS } from "../constants/ai";
import type {
  AIProvider,
  AIModel,
  ChatRequest,
  ChatResponse,
  FinishReason,
  GenerationOptions,
  ProviderType,
} from "../types/ai";

export type AIErrorCode = "invalid_api_key" | "rate_limited" | "network_error" | "unknown";

export class AIProviderError extends Error {
  constructor(
    message: string,
    public readonly code: AIErrorCode,
    public readonly provider: ProviderType,
  ) {
    super(message);
    this.name = "AIProviderError";
  }
}

function mapFinishReason(reason: string | null): FinishReason {
  switch (reason) {
    case "stop":
      return "stop";
    case "length":
      return "length";
    case "tool_calls":
    case "function_call":
      return "stop";
    default:
      return "error";
  }
}

const MODELS: AIModel[] = [
  {
    id: "gpt-4.1",
    provider: "openai",
    name: "GPT-4.1",
    capabilities: ["chat"],
    description: "OpenAI GPT-4.1 model for complex reasoning and instruction following.",
    maxTokens: 32768,
    supportsStreaming: false,
  },
  {
    id: "gpt-4.1-mini",
    provider: "openai",
    name: "GPT-4.1 Mini",
    capabilities: ["chat"],
    description: "Faster and more cost-effective variant of GPT-4.1.",
    maxTokens: 32768,
    supportsStreaming: false,
  },
  {
    id: "gpt-4o-mini",
    provider: "openai",
    name: "GPT-4o Mini",
    capabilities: ["chat"],
    description: "Fast, affordable small model for simple tasks.",
    maxTokens: 16384,
    supportsStreaming: false,
  },
];

export class OpenAIProvider implements AIProvider {
  readonly type: ProviderType = "openai";
  readonly name = "OpenAI";
  private client: OpenAI;

  constructor() {
    const apiKey = import.meta.env.VITE_OPENAI_API_KEY;
    this.client = new OpenAI({
      apiKey,
      dangerouslyAllowBrowser: true,
      maxRetries: 0,
    });
  }

  async generate(
    request: ChatRequest,
    options?: GenerationOptions,
  ): Promise<ChatResponse> {
    const resolvedOptions = { ...DEFAULT_GENERATION_OPTIONS, ...options };

    try {
      const response = await this.client.chat.completions.create({
        model: request.model,
        messages: request.messages.map((m) => ({
          role: m.role,
          content: m.content,
        })),
        temperature: resolvedOptions.temperature,
        max_tokens: resolvedOptions.maxTokens,
        top_p: resolvedOptions.topP,
        frequency_penalty: resolvedOptions.frequencyPenalty,
        presence_penalty: resolvedOptions.presencePenalty,
        stop: resolvedOptions.stop ?? undefined,
      });

      const choice = response.choices[0];
      const messageContent = choice.message.content ?? "";

      return {
        id: response.id,
        model: response.model,
        provider: "openai",
        message: {
          role: "assistant",
          content: messageContent,
        },
        usage: {
          promptTokens: response.usage?.prompt_tokens ?? 0,
          completionTokens: response.usage?.completion_tokens ?? 0,
          totalTokens: response.usage?.total_tokens ?? 0,
        },
        finishReason: mapFinishReason(choice.finish_reason),
      };
    } catch (error) {
      throw this.mapError(error);
    }
  }

  stream(
    _request: ChatRequest,
    _options?: GenerationOptions,
  ): AsyncIterable<ChatResponse> {
    throw new AIProviderError(
      "Streaming is not yet implemented for the OpenAI provider.",
      "unknown",
      "openai",
    );
  }

  async listModels(): Promise<AIModel[]> {
    return MODELS;
  }

  async healthCheck(): Promise<boolean> {
    const apiKey = import.meta.env.VITE_OPENAI_API_KEY;
    if (!apiKey) return false;
    try {
      this.client = new OpenAI({
        apiKey,
        dangerouslyAllowBrowser: true,
        maxRetries: 0,
      });
      return true;
    } catch {
      return false;
    }
  }

  private mapError(error: unknown): AIProviderError {
    if (error instanceof AuthenticationError) {
      return new AIProviderError(
        "Invalid API key. Check your VITE_OPENAI_API_KEY environment variable.",
        "invalid_api_key",
        "openai",
      );
    }

    if (error instanceof RateLimitError) {
      return new AIProviderError(
        "OpenAI rate limit exceeded. Please wait and try again.",
        "rate_limited",
        "openai",
      );
    }

    if (error instanceof APIConnectionError) {
      return new AIProviderError(
        "Failed to connect to OpenAI. Check your network connection.",
        "network_error",
        "openai",
      );
    }

    if (error instanceof APIError) {
      return new AIProviderError(
        error.message || "OpenAI API returned an error.",
        "unknown",
        "openai",
      );
    }

    return new AIProviderError(
      "An unexpected error occurred while calling OpenAI.",
      "unknown",
      "openai",
    );
  }
}

File: backend/app/services/ai/openai_provider.py
"""OpenAI provider implementation using the official Python SDK."""

import logging

import openai
from openai import (
    APIConnectionError,
    APIError,
    AuthenticationError,
    RateLimitError,
)

from app.config.settings import get_settings
from app.services.ai.base_provider import AIProvider
from app.services.ai.errors import AIAuthenticationError, AINetworkError, AIProviderError, AIRateLimitError
from app.services.ai.schemas import AIModel, ChatRequest, ChatResponse, FinishReason, ProviderType, TokenUsage

logger = logging.getLogger(__name__)

MODELS: list[AIModel] = [
    AIModel(
        id="gpt-4.1",
        provider="openai",
        name="GPT-4.1",
        capabilities=["chat", "streaming"],
        description="OpenAI GPT-4.1 model for complex reasoning and instruction following.",
        max_tokens=32768,
        supports_streaming=True,
    ),
    AIModel(
        id="gpt-4.1-mini",
        provider="openai",
        name="GPT-4.1 Mini",
        capabilities=["chat", "streaming"],
        description="Faster and more cost-effective variant of GPT-4.1.",
        max_tokens=32768,
        supports_streaming=True,
    ),
    AIModel(
        id="gpt-4o-mini",
        provider="openai",
        name="GPT-4o Mini",
        capabilities=["chat", "streaming"],
        description="Fast, affordable small model for simple tasks.",
        max_tokens=16384,
        supports_streaming=True,
    ),
]


def _map_finish_reason(reason: str | None) -> FinishReason:
    match reason:
        case "stop":
            return "stop"
        case "length":
            return "length"
        case "tool_calls" | "function_call":
            return "stop"
        case _:
            return "error"


class OpenAIProvider(AIProvider):
    """Provider implementation for OpenAI's chat completion API."""

    @property
    def type(self) -> ProviderType:
        return "openai"

    @property
    def name(self) -> str:
        return "OpenAI"

    def __init__(self) -> None:
        settings = get_settings()
        self._client = openai.AsyncOpenAI(
            api_key=settings.OPENAI_API_KEY,
            max_retries=0,
        )

    async def generate(self, request: ChatRequest) -> ChatResponse:
        try:
            response = await self._client.chat.completions.create(
                model=request.model,
                messages=[{"role": m.role, "content": m.content} for m in request.messages],
                temperature=request.temperature,
                max_tokens=request.max_tokens,
                top_p=request.top_p,
                frequency_penalty=request.frequency_penalty,
                presence_penalty=request.presence_penalty,
                stop=request.stop,
            )

            choice = response.choices[0]
            content = choice.message.content or ""

            return ChatResponse(
                id=response.id,
                model=response.model,
                provider="openai",
                message={"role": "assistant", "content": content},
                usage=TokenUsage(
                    prompt_tokens=response.usage.prompt_tokens if response.usage else 0,
                    completion_tokens=response.usage.completion_tokens if response.usage else 0,
                    total_tokens=response.usage.total_tokens if response.usage else 0,
                ),
                finish_reason=_map_finish_reason(choice.finish_reason),
            )
        except Exception as exc:
            raise self._map_error(exc)

    async def stream(self, request: ChatRequest):
        try:
            stream = await self._client.chat.completions.create(
                model=request.model,
                messages=[{"role": m.role, "content": m.content} for m in request.messages],
                temperature=request.temperature,
                max_tokens=request.max_tokens,
                top_p=request.top_p,
                frequency_penalty=request.frequency_penalty,
                presence_penalty=request.presence_penalty,
                stop=request.stop,
                stream=True,
            )

            async for chunk in stream:
                choice = chunk.choices[0] if chunk.choices else None
                if not choice:
                    continue

                content = choice.delta.content or ""
                finish_reason = _map_finish_reason(choice.finish_reason) if choice.finish_reason else None

                yield content, finish_reason

        except Exception as exc:
            raise self._map_error(exc)

    async def list_models(self) -> list[AIModel]:
        return MODELS

    async def health_check(self) -> bool:
        settings = get_settings()
        if not settings.OPENAI_API_KEY:
            return False
        try:
            await self._client.models.list()
            return True
        except Exception:
            return False

    def _map_error(self, error: Exception) -> AIProviderError:
        if isinstance(error, AuthenticationError):
            return AIAuthenticationError(provider="openai")
        if isinstance(error, RateLimitError):
            return AIRateLimitError(provider="openai")
        if isinstance(error, APIConnectionError):
            return AINetworkError(provider="openai")
        if isinstance(error, APIError):
            return AIProviderError(
                message=error.message or "OpenAI API returned an error.",
                code="unknown",
                provider="openai",
            )
        return AIProviderError(
            message="An unexpected error occurred while calling OpenAI.",
            code="unknown",
            provider="openai",
        )
